File: Circles.Profiles.Interfaces/IIpfsStore.cs
namespace Circles.Profiles.Interfaces;

/// <summary>Pure IPFS store abstraction.</summary>
public interface IIpfsStore
{
    /* Pins UTF-8 JSON (convenience) */
    Task<string> AddJsonAsync(string json,
        bool pin = true,
        CancellationToken ct = default);

    /* Raw byte variant – same semantics */
    Task<string> AddBytesAsync(ReadOnlyMemory<byte> bytes,
        bool pin = true,
        CancellationToken ct = default);

    Task<Stream> CatAsync(string cid, CancellationToken ct = default);

    Task<string> CatStringAsync(string cid, CancellationToken ct = default);

    /* hash-only: no upload */
    Task<string> CalcCidAsync(ReadOnlyMemory<byte> bytes,
        CancellationToken ct = default);
}

File: js/src/interfaces/IIpfsStore.ts
/**
 * Interface for interacting with IPFS.
 */
export interface IIpfsStore {
  /**
   * Adds JSON content to IPFS.
   * 
   * @param json JSON content to add
   * @param pin Whether to pin the content
   * @returns The CID of the added content
   */
  addJsonAsync(json: string, pin?: boolean): Promise<string>;
  
  /**
   * Adds binary data to IPFS.
   * 
   * @param bytes Binary data to add
   * @param pin Whether to pin the content
   * @returns The CID of the added content
   */
  addBytesAsync(bytes: Uint8Array, pin?: boolean): Promise<string>;
  
  /**
   * Adds content to IPFS (generic method)
   * @param content The content to add
   * @returns The CID of the added content
   */
  addAsync(content: Uint8Array | string): Promise<string>;
  
  /**
   * Retrieves content from IPFS as a byte array
   * @param cid The CID to retrieve
   * @returns The content as a byte array
   */
  getAsync(cid: string): Promise<Uint8Array>;
  
  /**
   * Pins content in IPFS to ensure persistence
   * @param cid The CID to pin
   */
  pinAsync(cid: string): Promise<void>;
  
  /**
   * Retrieves content from IPFS.
   * 
   * @param cid The CID to retrieve
   * @returns A stream of the content
   */
  catAsync(cid: string): Promise<ReadableStream<Uint8Array>>;
  
  /**
   * Retrieves content from IPFS as a string.
   * 
   * @param cid The CID to retrieve
   * @returns The content as a string
   */
  catStringAsync(cid: string): Promise<string>;
  
  /**
   * Calculates the CID of the given data.
   * 
   * @param bytes The data to calculate the CID for
   * @returns The calculated CID
   */
  calcCidAsync(bytes: Uint8Array): Promise<string>;
  
  /**
   * Disposes of resources.
   */
  dispose(): void;
}